fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java

```diff
@@ -107,40 +102,58 @@ private static long fileLockValueToTimestamp(byte[] value) {
                    Tuple.fromBytes(value).getLong(0);
         }
 
-        public void fileLockSet(long timeStampMillis) {
-            byte[] value = fileLockValue(timeStampMillis);
-            agilityContext.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_FILE_LOCK_SET,
-                    agilityContext.apply(aContext ->
-                            aContext.ensureActive().get(fileLockKey)
-                                    .thenAccept(val -> {
-                                        if (val != null) {
-                                            long existingTimeStamp =  fileLockValueToTimestamp(val);
-                                            if (existingTimeStamp > (timeStampMillis - timeWindowMilliseconds) &&
-                                                    existingTimeStamp < (timeStampMillis + timeWindowMilliseconds)) {
-                                                // Here: this lock is valid
-                                                throw new RecordCoreException("FileLock: Set: found old lock")
-                                                        .addLogInfo(LuceneLogMessageKeys.LOCK_EXISTING_TIMESTAMP, existingTimeStamp,
-                                                                LuceneLogMessageKeys.LOCK_DIRECTORY, this);
-                                            }
-                                            // Here: this lock is either too old, or in the future. Steal it
-                                            if (LOGGER.isWarnEnabled()) {
-                                                LOGGER.warn(KeyValueLogMessage.of("FileLock: Set: found old lock, discard it",
-                                                        LuceneLogMessageKeys.LOCK_EXISTING_TIMESTAMP, existingTimeStamp,
-                                                        LuceneLogMessageKeys.LOCK_DIRECTORY, this));
+        public void fileLockSet(boolean isHeartbeat) {
+            synchronized (fileLockSetLock) {
```

Comment:
This could cause issues with parallel threads being consumed waiting for the synchronized, but unable to actually start the asynchronous handling below, which could cause deadlocks in the reading, since you are synchronizing around the set. 
Can we pull the get out of the `synchronized` block.
You may have to store `this.timeStampMillis` in a temporary variable, and check that it didn't change while you did the read.

Reply:
I'm not sure how it can be done without including the `get' in a synchronized block. 
For example: 
```
myTimeStamp = this.timeStampMillis;  <<== local timeStampMillis at time A
agilityContext.asyncToSync(...
                        agilityContext.apply(aContext ->
                                aContext.ensureActive().get(fileLockKey) <<== reads remote timeStampMillis at time B
```
Or 
```
agilityContext.asyncToSync(...
                        agilityContext.apply(aContext ->
                                aContext.ensureActive().get(fileLockKey) <<== reads remote timeStampMillis at time A
                                        .thenAccept(val -> {
                      myTimeStamp = this.timeStampMillis;  <<== local timeStampMillis at time B
```
Is there an atomic `get(key,  lambdaToBeExecutedAtTheTimeOfTheGet)` operation?

Reply:
I might be missing something, but isn't it is safer to raise a semaphore before starting to wait on a worker thread via `asyncToSync`?

Reply:
There is not an atomic `get(key, lambdaToBeExecutedAtTheTimeOfTheGet)`

The problem is when this code is running in a `ForkJoinPool` with a parallelism of, e.g. 16. If you have 16 of these happening at the same time, you will have 16 threads inside the synchronized, waiting on a join. This has consumed all of the parallelism for the pool. So now, when the `get` goes to complete, there won't be any threads left on the `ForkJoinPool` to finish the `get`, this means that none of the `synchronized` blocks will ever complete, and the pool will be locked. 

This was the root cause behind #2234 

I think you can move the `synchronized` block after the `get`.

```java
        public void fileLockSet(boolean isHeartbeat) {
            agilityContext.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_FILE_LOCK_SET,
                    agilityContext.apply(aContext -> fileLockSet(isHeartbeat, aContext)
                    ));
        }

        private CompletableFuture<Void> fileLockSet(boolean isHeartbeat, FDBRecordContext aContext) {
            final long nowMillis = System.currentTimeMillis();
            return aContext.ensureActive().get(fileLockKey)
                    .thenAccept(val -> {
                        synchronized (fileLockSetLock) {
                            long existingTimeStamp = fileLockValueToTimestamp(val);
                            if (isHeartbeat) {
                                fileLockCheckHeartBeat(existingTimeStamp);
                            } else {
                                fileLockCheckNewLock(existingTimeStamp, nowMillis);
                            }
                            this.timeStampMillis = nowMillis;
                            byte[] value = fileLockValue(timeStampMillis);
                            aContext.ensureActive().set(fileLockKey, value);
                        }
                    });
        }
```

The fact that you are reading, and setting in the same `FDBRecordContext` protects you from the possibility that another transaction changed the timestamp.
I think you do need some handling to protect against the possibility of:
1. get -> t0
2. other -> updates to t1
3. existingTimeStamp now equals this.timeStampMillis which messes with the heartbeat code.

One simple solution would be to have the file lock contain a timestamp and a uuid. If the uuid changes you've lost the lock. Then the code in `fileLockSet` would validate that the uuid hasn't changed, and _only_ set a new timestamp if `nowMillis` is greater than `timeStampMillis` and the one that it read. Although, realistically, if it is the same uuid, and it goes down, it shouldn't be a problem.